research/conditional/moe_layers/load_balancing_loss.py

```diff
@@ -0,0 +1,28 @@
+import torch
+from fancy_einsum import einsum
+
+
+def calculate_load_balancing_loss(
+    alpha: float,
+    softmax_per_token: torch.Tensor,
+    n_tokens_in_each_expert: torch.Tensor,
+    use_einsum: bool = False,
+):
+    """
+    Calculates the load balancing loss for the token choice layer.
+
+    :param str alpha: aux loss weigth parameter
+    :param torch.Tensor softmax_per_token: tensor of shape (tokens, n_experts)
+    :param torch.Tensor tokens_in_each_expert: tensor of shape (n_experts)
+    """
+    n_tokens, n_experts = softmax_per_token.shape
+    assert n_experts == n_tokens_in_each_expert.shape[0]
+
+    per_expert_softmax_sum = torch.mean(softmax_per_token, dim=0)
```

Comment:
No clue, this is just copied from the previous implementation. Since it's a refactor I think it should be left as is (but maybe investigate it later?).